File: TimePickerTuto/app/src/main/java/com/example/timepickertuto/MainActivity.java
package com.example.timepickertuto;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;
import android.widget.TimePicker;

public class MainActivity extends AppCompatActivity {

    //필요한 변수 선언
    private TimePicker timePicker;
    private TextView time;
    private String format = "";
    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // 각각의 View를 변수에 할당
        timePicker = (TimePicker) findViewById(R.id.timePicker);
        time = (TextView) findViewById(R.id.time);
        // TimePicker의 시간과 분을 가져와서 변수에 할당
        int hour = timePicker.getHour();
        int min = timePicker.getMinute();

        // TimePicker View의 시간을 새로 맞출 시에 불리는 리스너 할당
        timePicker.setOnTimeChangedListener(new TimePicker.OnTimeChangedListener() {
            @Override
            public void onTimeChanged(TimePicker timePicker, int i, int i1) {
                showTime(timePicker.getHour(), timePicker.getMinute());//새로운 시간과 분을 text뷰에 띄워줌
            }
        });
        showTime(hour, min);//어플 시작 시 시간과 분을 text뷰에 띄워줌
    }

    public void showTime(int hour, int min){
        //시간에 따라 TextView에 표시할 값을 정함
        if (hour == 0){
            hour += 12;
            format = "AM";
        } else if (hour == 12){
            format = "PM";
        } else if (hour > 12){
            hour -= 12;
            format = "PM";
        } else  {
            format = "AM";
        }
        //정한 값을 새롭게 TextView에 표시
        time.setText(new StringBuilder().append(hour).append(":").append(min).append(" ").append(format));

    }
}

File: EditTextTuto/app/src/main/java/com/example/edittexttuto/MainActivity.kt
package com.example.edittexttuto

import android.os.Build
import android.os.Bundle
import android.os.CountDownTimer
import android.text.Spannable
import android.text.style.BackgroundColorSpan
import android.util.Log
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity


class MainActivity : AppCompatActivity() {
    lateinit var timer : CountDownTimer
    lateinit var tv : TextView
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val lyric =
            "[00:16:200]we wish you a merry christmas\n[00:18:300]we wish you a merry christmas\n[00:21:100]we wish you a merry christmas\n[00:23:600]and a happy new year\n"
        val list = lyric.split('[')
        var tmp = ""
        var i = 0
        var lengthStack = 0

        tv = findViewById(R.id.lyricView)
        tv.setText(lyric, TextView.BufferType.SPANNABLE)
        timer = object : CountDownTimer(10000, 500) {
            @RequiresApi(Build.VERSION_CODES.M)
            override fun onTick(millisUntilFinished: Long) {
                tv.text = setHigh(lengthStack, list[i].length)
                Log.i("tier", "$lengthStack ${lengthStack + list[i].length}")

                lengthStack += list[i].length
                if (i < list.lastIndex)
                    i++
                else {
                    i = 0
                    lengthStack = 0
                }

            }

            override fun onFinish() {
            }

        }.start()
    }

    @RequiresApi(Build.VERSION_CODES.M)
    private fun setHigh(start:Int, target: Int): Spannable {
        val spanLyric = tv.text as Spannable

        spanLyric.setSpan(
            BackgroundColorSpan(getColor(R.color.purple_500)), start, start + target,
            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
        )
        return spanLyric

    }
}